### ISSUE TITLE
Getting too many notifications with `dunst` crashes River

### ISSUE THREAD

**Issue by sagehane** [riverwm/river#555]
This issue probably stems from `dunst` (and the config), but I wasn't sure if River is supposed to crash in such a scenario, so I decided to report it.

How to reproduce:
1. Setup `dunst` with the "default" config: https://github.com/dunst-project/dunst/blob/master/dunstrc
2. Get a lot of notifications to display at once
3. River will crash

Sample script to cause the crash (value given to `seq` might have to be higher depending on screen/resolution?):
```sh
#!/bin/sh

for x in $(seq 30)
do
	notify-send "$x"
done
```

Reproduced on NixOS with version 0.1.3 and https://github.com/riverwm/river/commit/238c39379d461a97619241e0daf0a437eb60bbc3 with version 1.8.0 of `dunst`.

Notes:
- I presume it crashes because `dunst` keeps displaying notifications until the screen runs of space vertically in order to display it?
- It would cause River to crash while I left my computer idle, as notifications didn't disappear while `swaylock` was active.
- `mako` doesn't crash because it seems to only display 5 messages at a given time, by default.

**Comment by novakne** [comment]
I can reproduce it, also with fnott but not everytime, although I need 100 notifications to crash

- dunst log:

```
warning(layer_shell): margins of layer surface 'notifications' are too large to be reasonably handled. Closing.
debug(layer_shell): layer surface 'notifications' destroyed
Segmentation fault at address 0xc9
/home/novakane/src/river/river/Subsurface.zig:68:32: 0x27336e in Subsurface.create (river)
    assert(wlr_subsurface.data == 0);
                               ^
/home/novakane/src/river/river/Subsurface.zig:88:26: 0x240b7b in Subsurface.handleExisting (river)
        Subsurface.create(subsurface, parent);
                         ^
/home/novakane/src/river/river/LayerSurface.zig:69:30: 0x226f0d in LayerSurface.init (river)
    Subsurface.handleExisting(wlr_layer_surface.surface, .{ .layer_surface = self });
                             ^
/home/novakane/src/river/river/Server.zig:235:19: 0x226a03 in Server.handleNewLayerSurface (river)
    node.data.init(output, wlr_layer_surface);
                  ^
/home/novakane/src/river/zig-cache/zig-wayland/wayland_server_core.zig:492:25: 0x226f90 in .wayland.wayland_server_core.struct:490:17.wrapper (river)
                        @call(.{ .modifier = .always_inline }, notify, .{ listener, @intToPtr(T, @ptrToInt(data)) });
                        ^
../util/signal.c:29:3: 0x7f56613a6e2b in wlr_signal_emit_safe (../util/signal.c)
../types/wlr_layer_shell_v1.c:365:3: 0x7f5661387784 in layer_surface_role_commit (../types/wlr_layer_shell_v1.c)
../types/wlr_surface.c:461:3: 0x7f5661397c55 in surface_commit_state (../types/wlr_surface.c)
../types/wlr_surface.c:562:3: 0x7f5661398009 in surface_handle_commit (../types/wlr_surface.c)
???:?:?: 0x7f5660f83e6c in ??? (???)
thread 1273 panic: integer overflow
/home/novakane/.local/lib/zig-0.9.1/lib/std/debug.zig:141:69: 0x2a3d52 in std.debug.dumpStackTraceFromBase (river)
            printSourceAtAddress(debug_info, stderr, return_address - 1, tty_config) catch return;
                                                                    ^
/home/novakane/.local/lib/zig-0.9.1/lib/std/debug.zig:1714:35: 0x281ac7 in std.debug.handleSegfaultPosix (river)
            dumpStackTraceFromBase(bp, ip);
                                  ^
???:?:?: 0x7f56612f58ff in ??? (../sysdeps/unix/sysv/linux/x86_64/sigaction.c)
/home/novakane/src/river/river/Subsurface.zig:88:26: 0x240b7b in Subsurface.handleExisting (river)
        Subsurface.create(subsurface, parent);
                         ^
/home/novakane/src/river/river/LayerSurface.zig:69:30: 0x226f0d in LayerSurface.init (river)
    Subsurface.handleExisting(wlr_layer_surface.surface, .{ .layer_surface = self });
                             ^
/home/novakane/src/river/river/Server.zig:235:19: 0x226a03 in Server.handleNewLayerSurface (river)
    node.data.init(output, wlr_layer_surface);
                  ^
/home/novakane/src/river/zig-cache/zig-wayland/wayland_server_core.zig:492:25: 0x226f90 in .wayland.wayland_server_core.struct:490:17.wrapper (river)
                        @call(.{ .modifier = .always_inline }, notify, .{ listener, @intToPtr(T, @ptrToInt(data)) });
                        ^
../util/signal.c:29:3: 0x7f56613a6e2b in wlr_signal_emit_safe (../util/signal.c)
../types/wlr_layer_shell_v1.c:365:3: 0x7f5661387784 in layer_surface_role_commit (../types/wlr_layer_shell_v1.c)
../types/wlr_surface.c:461:3: 0x7f5661397c55 in surface_commit_state (../types/wlr_surface.c)
../types/wlr_surface.c:562:3: 0x7f5661398009 in surface_handle_commit (../types/wlr_surface.c)
???:?:?: 0x7f5660f83e6c in ??? (???)
???:?:?: 0x0 in ??? (???)
(EE) failed to read Wayland events: Broken pipe
dbus-daemon[1272]: [session uid=1000 pid=1272] Activating service name='org.freedesktop.Notifications' requested by ':1.145' (uid=1000 pid=3039 comm="notify-send 33 ")
A connection to the bus can't be made
WARNING: Cannot open X11 display.
ERROR: [  get_x11_output:0065] Couldn't initialize X11 output. Aborting...
```

- fnott log

```
warning(layer_shell): margins of layer surface 'notifications' are too large to be reasonably handled. Closing.
debug(layer_shell): layer surface 'notifications' unmapped
debug(layer_shell): layer surface 'notifications' destroyed
Segmentation fault at address 0x8
/home/novakane/src/river/river/LayerSurface.zig:167:23: 0x240b00 in LayerSurface.handleCommit (river)
    self.output.damage.addWhole();
                      ^
/home/novakane/src/river/zig-cache/zig-wayland/wayland_server_core.zig:492:25: 0x2408e0 in .wayland.wayland_server_core.struct:490:17.wrapper (river)
                        @call(.{ .modifier = .always_inline }, notify, .{ listener, @intToPtr(T, @ptrToInt(data)) });
                        ^
../util/signal.c:29:3: 0x7fb154594e2b in wlr_signal_emit_safe (../util/signal.c)
../types/wlr_surface.c:464:2: 0x7fb154585c6f in surface_commit_state (../types/wlr_surface.c)
../types/wlr_surface.c:562:3: 0x7fb154586009 in surface_handle_commit (../types/wlr_surface.c)
???:?:?: 0x7fb154171e6c in ??? (???)
thread 1275 panic: integer overflow
/home/novakane/.local/lib/zig-0.9.1/lib/std/debug.zig:141:69: 0x2a3d52 in std.debug.dumpStackTraceFromBase (river)
            printSourceAtAddress(debug_info, stderr, return_address - 1, tty_config) catch return;
                                                                    ^
/home/novakane/.local/lib/zig-0.9.1/lib/std/debug.zig:1714:35: 0x281ac7 in std.debug.handleSegfaultPosix (river)
            dumpStackTraceFromBase(bp, ip);
                                  ^
???:?:?: 0x7fb1544e38ff in ??? (../sysdeps/unix/sysv/linux/x86_64/sigaction.c)
/home/novakane/src/river/zig-cache/zig-wayland/wayland_server_core.zig:492:25: 0x2408e0 in .wayland.wayland_server_core.struct:490:17.wrapper (river)
                        @call(.{ .modifier = .always_inline }, notify, .{ listener, @intToPtr(T, @ptrToInt(data)) });
                        ^
../util/signal.c:29:3: 0x7fb154594e2b in wlr_signal_emit_safe (../util/signal.c)
../types/wlr_surface.c:464:2: 0x7fb154585c6f in surface_commit_state (../types/wlr_surface.c)
../types/wlr_surface.c:562:3: 0x7fb154586009 in surface_handle_commit (../types/wlr_surface.c)
???:?:?: 0x7fb154171e6c in ??? (???)
???:?:?: 0x0 in ??? (???)
(EE) failed to read Wayland events: Broken pipe
```

**Comment by novakne** [comment]
Get a gdb log too

```log
Starting program: /home/novakane/.local/bin/river -log-level debug
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
[New Thread 0x7fffee856640 (LWP 3415)]
[New Thread 0x7fffed855640 (LWP 3416)]
[New Thread 0x7fffe7fff640 (LWP 3417)]
[New Thread 0x7fffe6ffe640 (LWP 3418)]
[New Thread 0x7fffe5ffd640 (LWP 3419)]
[New Thread 0x7fffd3fff640 (LWP 3420)]
[New Thread 0x7fffd2ffe640 (LWP 3421)]
[New Thread 0x7fffd1ffd640 (LWP 3422)]
[New Thread 0x7fffc3fff640 (LWP 3423)]
[Detaching after fork from child process 3424]
[Detaching after fork from child process 3426]
[Detaching after fork from child process 3437]
[Detaching after fork from child process 3446]
[Detaching after fork from child process 3594]
[Detaching after fork from child process 3691]

Thread 1 "river" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
49	../sysdeps/unix/sysv/linux/raise.c: Aucun fichier ou dossier de ce type.
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
        set = {__val = {16386, 0, 1182793728, 0, 1110225085, 0, 1106515194, 0, 140737488333569, 12920317831840530434, 0, 0, 2, 9223372036854775822, 0, 0}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007ffff7b9a536 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x26bae6 <std.dwarf.Die.getAttrSecOffset+86>, sa_sigaction = 0x26bae6 <std.dwarf.Die.getAttrSecOffset+86>}, sa_mask = {__val = {2538214, 2538214, 2538214, 
              2538214, 2538214, 2538214, 2538214, 2538214, 2538214, 3, 140737354129824, 157, 2342837, 2538214, 0, 3073440}}, sa_flags = 2538214, sa_restorer = 0x3}
        sigs = {__val = {32, 2, 775, 140737488333992, 32, 2520838, 2444884, 2447925, 2444147, 2440928, 2443341, 2440928, 2538214, 2538214, 2538214, 2538214}}
#2  0x0000000000220239 in std.os.abort () at /home/novakane/.local/lib/zig-0.9.1/lib/std/os.zig:443
No locals.
#3  0x000000000021b4bb in std.debug.panicImpl (trace=0x0, first_trace_addr=..., msg=...) at /home/novakane/.local/lib/zig-0.9.1/lib/std/debug.zig:329
No locals.
#4  0x000000000021a1e8 in std.builtin.default_panic (msg=<error reading variable: Cannot access memory at address 0x0>, error_return_trace=0x0) at /home/novakane/.local/lib/zig-0.9.1/lib/std/builtin.zig:744
        first_trace_addr = 2569762
#5  0x0000000000273622 in .wayland.wayland_server_core.list.Head(.wlroots.types.surface.ParentState,"link").Iterator(.wayland.wayland_server_core.Direction.forward).next (it=0x7fffffffb038)
    at /home/novakane/src/river/zig-cache/zig-wayland/wayland_server_core.zig:414
No locals.
#6  0x0000000000240b51 in Subsurface.handleExisting (surface=0xac1929, parent=...) at Subsurface.zig:86
        below_it = {head = 0xac1a49, current = 0xac1a49}
        parent_state = 0xac5358
        above_it = {head = 0xac5358, current = 0x9a6e78}
        parent_state = 0x7fffffffb030
#7  0x0000000000226f0e in LayerSurface.init (self=0xac52d0, output=0x9a6e20, wlr_layer_surface=0xb49310) at LayerSurface.zig:69
No locals.
#8  0x0000000000226a04 in Server.handleNewLayerSurface (listener=0x2ee008 <server+96>, wlr_layer_surface=0xb49310) at Server.zig:235
        self = 0x2edfa8 <server>
        node = 0xac52c0
        output = 0x9a6e20
#9  0x0000000000226f91 in .wayland.wayland_server_core.struct:490:17.wrapper (listener=0x2ee008 <server+96>, data=0xb49310) at /home/novakane/src/river/zig-cache/zig-wayland/wayland_server_core.zig:492
No locals.
#10 0x00007ffff7dfee2c in wlr_signal_emit_safe (signal=0x96dcd0, data=0xb49310) at ../util/signal.c:29
        pos = 0x2ee008 <server+96>
        l = 0x2ee008 <server+96>
        cursor = {link = {prev = 0x2ee008 <server+96>, next = 0x7fffffffb320}, notify = 0x7ffff7dfed76 <handle_noop>}
        end = {link = {prev = 0x7fffffffb300, next = 0x96dcd0}, notify = 0x7ffff7dfed76 <handle_noop>}
#11 0x00007ffff7ddf785 in layer_surface_role_commit (wlr_surface=0xac2ef0) at ../types/wlr_layer_shell_v1.c:365
        surface = 0xb49310
        horiz = 12
        vert = 3
        __PRETTY_FUNCTION__ = "layer_surface_role_commit"
#12 0x00007ffff7defc56 in surface_commit_state (surface=0xac2ef0, next=0xac3078) at ../types/wlr_surface.c:461
        __PRETTY_FUNCTION__ = "surface_commit_state"
        invalid_buffer = false
        subsurface = 0xac30e0
#13 0x00007ffff7df000a in surface_handle_commit (client=0xa25d60, resource=0xac5b40) at ../types/wlr_surface.c:562
        surface = 0xac2ef0
        subsurface = 0x0
#14 0x00007ffff79dbe6d in ?? () from /usr/lib/libffi.so.7
No symbol table info available.
#15 0x00007ffff79db2aa in ?? () from /usr/lib/libffi.so.7
No symbol table info available.
#16 0x00007ffff7f4c085 in ?? () from /lib64/libwayland-server.so.0
No symbol table info available.
#17 0x00007ffff7f500bc in ?? () from /lib64/libwayland-server.so.0
No symbol table info available.
#18 0x00007ffff7f4ec12 in wl_event_loop_dispatch () from /lib64/libwayland-server.so.0
No symbol table info available.
#19 0x00007ffff7f4f365 in wl_display_run () from /lib64/libwayland-server.so.0
No symbol table info available.
#20 0x000000000021c4d9 in main () at main.zig:126
        cmd = {ptr = 0x310eb0 "\030", len = 33}
        child_pgid = {val = 3426, maybe = true}
        pgid = 0
        err = WouldBlock
        argv = {ptr = 0x7fffffffc1d8, len = 3}
        result = {args = {ptr = 0x7fffffffc1f0, len = 0}, flag_data = {{name = 0x216817 <__unnamed_14> "-h", value = {payload = {boolean = false, arg = 0x0}, tag = boolean}}, {
              name = 0x21681a <__unnamed_16> "-version", value = {payload = {boolean = false, arg = 0x0}, tag = boolean}}, {name = 0x216823 <__unnamed_20> "-c", value = {payload = {boolean = false, 
                  arg = 0x0}, tag = arg}}, {name = 0x216826 <__unnamed_18> "-log-level", value = {payload = {boolean = false, arg = 0x7fffffffc5f6 "debug"}, tag = arg}}}}
        level = {ptr = 0x7fffffffc5f6 "debug", len = 5}
        startup_command = {val = {ptr = 0x310eb0 "\030", len = 33}, maybe = true}
#21 0x00000000002521d8 in std.start.callMain () at /home/novakane/.local/lib/zig-0.9.1/lib/std/start.zig:561
        err = AddSocketFailed
#22 0x0000000000220078 in std.start.initEventLoopAndCallMain () at /home/novakane/.local/lib/zig-0.9.1/lib/std/start.zig:495
No locals.
#23 std.start.callMainWithArgs (argc=3, argv=0x7fffffffc1d8, envp=...) at /home/novakane/.local/lib/zig-0.9.1/lib/std/start.zig:445
No locals.
#24 0x000000000021fe23 in std.start.main (c_argc=3, c_argv=0x7fffffffc1d8, c_envp=0x7fffffffc1f8) at /home/novakane/.local/lib/zig-0.9.1/lib/std/start.zig:460
        env_count = 75
        envp = {ptr = 0x7fffffffc1f8, len = 75}

```

**Comment by ifreund** [comment]
Should be fixed by adf6e1f19db41dc7e722fb00da1a7a350f7275ed, thanks for the report and stack traces!

**Comment by ifreund** [comment]
> This issue probably stems from dunst (and the config), but I wasn't sure if River is supposed to crash in such a scenario, so I decided to report it.

Also to be clear, river should never crash no matter what crazy things clients do. Every river crash is a bug.
